### PR TITLE
Preserve Snowbridge v1 inbound liveness when the Asset Hub sovereign is empty

### PR DESCRIPTION
**Pull request by alistair-singh**
# Description

Snowbridge v1 inbound messages now burn the teleported Asset Hub execution fee from the relayer's Bridge Hub account instead of requiring funds in the Asset Hub sovereign account on Bridge Hub. The Asset Hub sovereign reimburses the relayer best-effort for the burned fee plus delivery reward when it has enough reducible balance.

This preserves the existing funded-sovereign economics while preventing an empty Asset Hub sovereign from stalling the Ethereum-to-Polkadot inbound channel. In the empty-sovereign case, the message is still processed and the nonce advances, but the relayer is not reimbursed.

## Review Notes

Motivation: In order to deprecate Snowbridge V1, we can simply allow funds to run out in these sovereigns making it unattractive to relayers. If the community wishes to continue to use V1, we can submit a proposal to top up these accounts  from treasury so that relayers are rewarded again or they can migrate to V2. In reality we will probably submit at least 1 proposal to top up these accounts one last time to give the community time to migrate over.

# Checklist

* [x] My PR includes a detailed description as outlined in the "Description" and its two subsections above.
* [x] My PR follows the [labeling requirements](
https://github.com/paritytech/polkadot-sdk/blob/master/docs/contributor/CONTRIBUTING.md#Process
) of this project (at minimum one label for `T` required)
    * External contributors: Use `/cmd label <label-name>` to add labels
    * Maintainers can also add labels manually
* [x] I have made corresponding changes to the documentation (if applicable)
* [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs [hunk 1]
```diff
@@ -349,10 +351,13 @@ pub mod pallet {
 		}
 
 		/// Burn the amount of the fee embedded into the XCM for teleports
-		pub fn burn_fees(para_id: ParaId, fee: BalanceOf<T>) -> DispatchResult {
+		pub fn burn_fees(para_id: ParaId, who: &T::AccountId, fee: BalanceOf<T>) -> DispatchResult {
 			let dummy_context =
 				XcmContext { origin: None, message_id: Default::default(), topic: None };
 			let dest = Location::new(1, [Parachain(para_id.into())]);
+			let relayer_id: [u8; 32] =
+				who.encode().try_into().map_err(|_| TokenError::FundsUnavailable)?;
```
Comment (claravanstaden) [inline bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: Not sure this error is correct?

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs [hunk 2]
```diff
@@ -295,7 +282,22 @@ pub mod pallet {
...
+			let delivery_cost = Self::calculate_delivery_cost(event.encode().len() as u32);
+			let owed = fee.saturating_add(delivery_cost);
+			let amount = T::Token::reducible_balance(
+				&sovereign_account,
+				Preservation::Preserve,
+				Fortitude::Polite,
+			)
+			.min(owed);
+			if !amount.is_zero() {
+				T::Token::transfer(&sovereign_account, &who, amount, Preservation::Preserve)?;
```
Comment (claravanstaden) [inline bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: If this fails, it will abort the whole method right? So not truly best effort? Is this for temporary errors that can be retried?

**Review by claravanstaden — APPROVED** [review]
Looks good! Should update the benchmarks too pls.

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs (hunk 2 above)
Comment (yrong) [reply bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: I assume this is fine since the amount is `min(reducible_balance, refund)`, which is guarded in the previous step, so it shouldn't be able to fail here. That said, it would be good to add a test case for this scenario—or perhaps it's not really an edge case, since we'll intentionally drain the sovereign account to verify that the relayer correctly covers the full cost. As a bonus, this should also allow all existing V1 and V2 test cases (in both directions) to continue passing even with an empty sovereign account.

**Comment by alistair-singh** [comment]
/cmd bench --runtime bridge-hub-westend --pallet snowbridge_pallet_inbound_queue

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs (hunk 1 above)
Comment (alistair-singh) [reply bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: fixed in [use correct error message](https://github.com/paritytech/polkadot-sdk/pull/12569/commits/aa043fc5fb01fed1cbce2137787e4c47ee15cee6)

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs (hunk 2 above)
Comment (alistair-singh) [reply bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: The old code did the same thing. Although ron is right that the min stops us from getting here, there is an edge case where dusted fees could cause revert, now this would mean the relayer balance is already extremely low anyway so i am not sure we need a fix. But added it anyway.

https://github.com/paritytech/polkadot-sdk/pull/12569/commits/980bc9005aca296a0be3d9fe59fcb7571eaf5f41

**Comment by alistair-singh** [comment]
/cmd bench --runtime bridge-hub-westend --pallet snowbridge_pallet_inbound_queue

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs (hunk 2 above)
Comment (alistair-singh) [reply bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: Ron, the `test_submit_no_funds_to_reward_relayers_just_ignore` unit test and `send_eth_asset_from_ethereum_with_empty_asset_hub_sovereign_preserves_liveness` integration test tests this. 

I have added some tests to test the dusted `test_submit_ignores_reimbursement_below_existential_deposit` and broke relayer cases `test_submit_fails_and_nonce_reverts_when_relayer_cannot_front_fee`.

https://github.com/paritytech/polkadot-sdk/pull/12569/commits/985fb47d65f25aab1f546d3ade4b16ae9947b6bd

**Review by yrong — APPROVED** [review]
+1

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs [hunk 3]
```diff
@@ -295,7 +282,34 @@ pub mod pallet {
...
+				// below the existential deposit) must never revert message processing, otherwise
+				// the nonce would roll back and stall the channel. The relayer simply goes
+				// unrewarded for this message.
+				if let Err(error) =
+					T::Token::transfer(&sovereign_account, &who, amount, Preservation::Preserve)
+				{
+					tracing::error!(
+						target: LOG_TARGET,
+						?error,
+						"Failed to reimburse relayer from sovereign account"
```
Comment (vgeddes) [inline bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: It's not useful to emit this log. This behavior is not an error and it's a valid code path.

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs [hunk 4]
```diff
@@ -295,7 +282,34 @@ pub mod pallet {
...
+			let delivery_cost = Self::calculate_delivery_cost(event.encode().len() as u32);
+			let owed = fee.saturating_add(delivery_cost);
+			let amount = T::Token::reducible_balance(
+				&sovereign_account,
+				Preservation::Preserve,
+				Fortitude::Polite,
+			)
+			.min(owed);
+			if !amount.is_zero() {
+				// Best-effort: a failed reimbursement (e.g. the transfer would dust either account
```
Comment (vgeddes) [inline bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: Quite a verbose comment. Can be shortened to: Best effort attempt to reward relayer if there are funds available.

**Review by vgeddes — CHANGES_REQUESTED** [review]

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs (hunk 3 above)
Comment (alistair-singh) [reply bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: Resovled in: https://github.com/paritytech/polkadot-sdk/pull/12569/commits/1e0b467ecafc3ca58e504ac4390735ffff3f6bfc

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs (hunk 4 above)
Comment (alistair-singh) [reply bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: Resovled in: https://github.com/paritytech/polkadot-sdk/pull/12569/commits/1e0b467ecafc3ca58e504ac4390735ffff3f6bfc

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs [hunk 5]
```diff
@@ -349,10 +363,13 @@ pub mod pallet {
 		}
 
 		/// Burn the amount of the fee embedded into the XCM for teleports
-		pub fn burn_fees(para_id: ParaId, fee: BalanceOf<T>) -> DispatchResult {
+		fn burn_fees(para_id: ParaId, who: &T::AccountId, fee: BalanceOf<T>) -> DispatchResult {
```
Comment (alistair-singh) [inline bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: I reviewed again. The only thing I can see in this code block that can be improved on is the account conversion from `AccountId` to `[u8; 32]`. We can directly take on the trait bound `trait Config: frame_system::Config<AccountId: Into<[u8; 32]>>` and just use `.into()` instead of using `.encode().try_into()`. The two code paths produce the same results for our use cases of 32 byte substrate accounts so I wouldn't call this a bug. Is the the bug you are talking about? Can you just list the bugs?

Comment (vgeddes) [reply bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: What is the role of `dest` here? And why are you checking out assets from dest instead of relayer?

Comment (vgeddes) [reply bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: And yeah, at least on a semantic level, `Into<[u8; 32]>` is more correct, even if both approaches lead to the same result.

Comment (vgeddes) [reply bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: Ignore my comment about `dest`, current code looks fine.

**Review by vgeddes — APPROVED** [review]

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs [hunk 6]
```diff
@@ -349,10 +353,13 @@ pub mod pallet {
...
```
Comment (vgeddes) [inline bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: A good opportunity to make the signature clearer

```
burn_for_teleport(dest, from, amount)
```

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs (hunk 5 above)
Comment (alistair-singh) [reply bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: > And yeah, at least on a semantic level, `Into<[u8; 32]>` is more correct, even if both approaches lead to the same result.

https://github.com/paritytech/polkadot-sdk/pull/12569/commits/68cb91e3d333d5a881dc7e18fda255a50e4a15cb

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs [hunk 7]
```diff
@@ -348,11 +351,17 @@ pub mod pallet {
 				.saturating_add(T::PricingParameters::get().rewards.local)
 		}
 
-		/// Burn the amount of the fee embedded into the XCM for teleports
-		pub fn burn_fees(para_id: ParaId, fee: BalanceOf<T>) -> DispatchResult {
+		/// Teleport the fee, fronted by the relayer, to the destination parachain.
+		fn burn_for_teleport(
+			para_id: ParaId,
```
Comment (vgeddes) [inline bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: nit(naming): 
```suggestion
			dest: ParaId,
```

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs [hunk 8]
```diff
@@ -348,11 +351,17 @@ pub mod pallet {
...
+			who: &T::AccountId,
+			fee: BalanceOf<T>,
```
Comment (vgeddes) [inline bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: nit: since `burn_for_teleport` is generalised, so should the naming:

```suggestion
			amount: BalanceOf<T>,
```

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs [hunk 9]
```diff
@@ -348,11 +351,17 @@ pub mod pallet {
...
+		) -> DispatchResult {
 			let dummy_context =
 				XcmContext { origin: None, message_id: Default::default(), topic: None };
 			let dest = Location::new(1, [Parachain(para_id.into())]);
+			let relayer_id: [u8; 32] = who.clone().into();
+			let relayer = Location::new(0, [AccountId32 { network: None, id: relayer_id }]);
```
Comment (vgeddes) [inline bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: nit: can just rebind `who` instead of using the name `relayer` which doesn't make meaningful sense within this generalised `burn_for_teleport` function.

```suggestion
			let who = who.clone().into();
			let who = Location::new(0, [AccountId32 { network: None, id: who }]);
```

**Comment by alistair-singh** [comment]
/cmd bench --runtime bridge-hub-westend --pallet snowbridge_pallet_inbound_queue

**Review by serban300 — APPROVED** [review]

**Review by rosarp — APPROVED** [review]

**Review by acatangiu — APPROVED** [review]

**Comment by acatangiu** [comment]
actually, wait a bit before merging, I need to double-check something

**Review by acatangiu — APPROVED** [review]
all good
